Add declarative macro guideline: avoid specializing (#16)
rustfoundation/safety-critical-rust-coding-guidelines: src/coding-guidelines/macros.rst

Macros
======

.. guideline:: Avoid specialized, fixed patterns within declarative macros
   :id: gui_FSpI084vbwmJ
   :status: draft
   :fls: fls_w44hav7mw3ao
   :tags: reduce-human-error
   :category: macros
   :recommendation: encouraged

>    Description of the guideline goes here.

Reviewer: ```suggestion
    Matchers within macro rules are evaluated sequentially and short-circuit on
    the first match. If a specialized fixed matcher follows a broader matcher,
    it may be unreachable. This can lead to subtle and surprising bugs. It is
    encouraged to avoid the use of specialized, fixed matchers.
```

   .. rationale::
      :id: rat_zqr9uEqP6nzW
      :status: draft

      It's common to use macros to avoid writing repetitive code, such as trait

>        implementations. It's possible to use derive macros or declarative macros
>        to do so.

Reviewer: ```suggestion
      implementations. It's possible to use derive macros or declarative macros
      to do so.
```

      In a declarative macro the ordering of the patterns will be the order that
      they are matched against which can lead to unexpected behavior in the case
      where we have unique behavior intended for a particular expression.

      If needing to specialize logic within the macro based on a particular
      expression's value, it may be better to not use a declarative macro.

   .. bad_example::
      :id: bad_ex_5vK0CCmePkef
      :status: draft

>       We have two macro match rules at the same level of nesting. Since the
>       matching is done macro rule by macro rule and this process is stopped as soon
>       as a macro matcher is matched we will not reach the specialized case for EmergencyValve.

Reviewer: ```suggestion
      We have two macro match rules at the same level of nesting. Since macro
      matching is done sequentially through the matchers and stops at the first 
      match, the specialized case for EmergencyValve is unreachable.
```

Just a suggested wording change -- feel free to take it if you like it better.

Macros
======

.. guideline:: Avoid specialized, fixed patterns within declarative macros
   :id: gui_FSpI084vbwmJ
   :status: draft
   :fls: fls_w44hav7mw3ao
   :tags: reduce-human-error
   :category: macros
   :recommendation: encouraged

   Description of the guideline goes here.

   .. rationale::
      :id: rat_zqr9uEqP6nzW
      :status: draft

      It is common to use declarative macros to implement traits which would

>       otherwise involve repetitive code.

Reviewer 2: I would say that a derive macro is the more idiomatic way of generating trait implementations, while a declarative macro is more suitable as a syntax extension mechanism as well as for generating boiler plate code outside of places where derive macros are usable (trait implementations can of course be considered as boiler plate at times). 

Perhaps we should phrase this so that we don't give the impression that generating trait implementations via `macro_rules!` is somehow a preferred or recommended way?

Author: Good point. I think that a derive macro will often be the more idiomatic thing. There is a bit of trade-off to be had here where that requires bringing in external crates (proc-macro2, syn, quote) which could make safety-qualification more challenging.

I think this gets into the discussion we had in the last coding guidelines meeting that Rust is a sophisticated language with a number of ways to do things.

I could see rephrasing this to the below to avoid biasing one way or another in this guideline:

> It's common to use macros to avoid writing repetitive code, such as trait implementations. It's possible to use derive macros or declarative macros to do so.

I think we could also have another guideline which said just what you pointed out that we'd recommend using a derive macro for repetitive trait implementation, if the additional external crates are not of issue. What do you think?

Author: Thanks for the feedback @manthonyaiello! I feel like we're in one of those "depends on your context" scenarios.

This coding guideline is setting out to suggest something for declarative macros specifically, which would be in libcore.

How do you feel about the text that's there currently more as background in this coding guideline?
> It's common to use macros to avoid writing repetitive code, such as trait implementations. It's possible to use derive macros or declarative macros to do so.

On another note, we could have another coding guideline we can give rationale on the circumstances:
* if you need safety-certification, probably stick to libcore and declarative macros for use cas of DRY
* if you are comfortable without safety-certification, derive macros can offer a bit more ergonomics

Reviewer 3: If the purpose of the coding guideline/standard is overall safety-critical use cases, then I think certification can/should be assumed and we should focus on those needs. I would lean away from recommending alternative approaches until/unless they can be met by a certifiable/certified runtime. Just my 2¢.

Reviewer 2: My point was not intended to advocate for one approach over another, but rather that we should avoid giving the impression that a non-idiomatic way is recommended. I know it was not the intention but the original text could maybe have been interpreted that way, was my concern.

We can discuss it at the meeting how to best approach these types of situations where the idiomatic way is perhaps more difficult to certify since that is obviously an important point. But just a note: neither declarative or derive macros impose any additional run-time components on their own since both are compile-time code generation mechanisms. But the latter method adds (in practice) additional "semi-standard" third party crates as Pete pointed out.

      In a declarative macro the ordering of the patterns will be the order that
      they are matched against which can lead to unexpected behavior in the case
      where we have unique behavior intended for a particular expression.

      If needing to specialize logic within the macro based on a particular
      expression's value, it may be better to not use a declarative macro.

      Limitation: Note that following this rule means that we are unable to support

>       variadic declarative macros with one or more arguments.

Reviewer 2: Unsure if if I follow, but doesn't the repeat syntax in `macro_rules!` allow for specifying handling for zero or more arguments as one rule? Its been a while since we wrote a macro so I am not 100% sure.

Author: Oh, you're totally right! Thanks for pointing this out.

I took [this example](https://doc.rust-lang.org/rust-by-example/macros/variadics.html) from Rust by Example and came up with [this example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=f606fc667995612b81b429c091135fc3) which shows that it's possible to support variadic declarative macros even with a single macro rule at a certain level of nesting.

Will update.

Author: Done!

Macros
======

.. guideline:: Avoid specialized, fixed patterns within declarative macros
   :id: gui_FSpI084vbwmJ
   :category: required
   :status: draft
   :fls: fls_w44hav7mw3ao
   :decidability: decidable
   :scope: module
   :tags: reduce-human-error

    Matchers within macro rules are evaluated sequentially and short-circuit on
    the first match. If a specialized fixed matcher follows a broader matcher,
    it may be unreachable. This can lead to subtle and surprising bugs. It is
    encouraged to avoid the use of specialized, fixed matchers.

   .. rationale::
      :id: rat_zqr9uEqP6nzW
      :status: draft

      It's common to use macros to avoid writing repetitive code, such as trait

>       implementations. It's possible to use derive macros or declarative macros

Reviewer 4: ```suggestion
      implementations. It's possible to use procedural macros or declarative macros
```
You can very well use attributes and functional-proc macros for this as well

      to do so.

      In a declarative macro the ordering of the patterns will be the order that
      they are matched against which can lead to unexpected behavior in the case

>       where we have unique behavior intended for a particular expression.

Reviewer 4: > In a declarative macro the ordering of the patterns will be the order that they are matched against

This doesn't really say anything. What is the order of the patterns? As in, what do they express. I assume you are intending to describe the selection process of which rule is selected by the compiler? I'd rewrite this paragraph (not quite sure how) as the current writing is rather confusing to me (and doesn't really say much)

Author: @Veykril -- you're right that this is the attempt made here to describe:
> I assume you are intending to describe the selection process of which rule is selected by the compiler? 

Could I have your help with a proposed wording here?

Reviewer 4: ```

The choice of which transcriber is being used in a declarative macro depends on their declaration order within the macro. This can lead to unexpected behavior changes in invocations of declarative macros if a new transciber is inserted before another due to invocations suddenly matching a different transcriber.
```

Author: TIL that part is called a transcriber!

>       If needing to specialize logic within the macro based on a particular
>       expression's value, it is better to not use a declarative macro.

Reviewer 4: ```suggestion
      If needing to specialize logic within the macro based on a particular
      expression's value, it is better to not use a declarative macro with multiple rules.
```

This change implies using distinct macros for the specialized job is fine (which imo it is)

Author: I agree 👍

Macros
======

.. guideline:: Avoid specialized, fixed patterns within declarative macros
   :id: gui_FSpI084vbwmJ
   :category: required
   :status: draft
   :release: 1.85.0;1.85.1
   :fls: fls_w44hav7mw3ao
   :decidability: decidable
   :scope: module
   :tags: reduce-human-error

    Matchers within macro rules are evaluated sequentially and short-circuit on
    the first match. If a specialized fixed matcher follows a broader matcher,
    it may be unreachable. This can lead to subtle and surprising bugs. It is
    encouraged to avoid the use of specialized, fixed matchers.

   .. rationale::
      :id: rat_zqr9uEqP6nzW
      :status: draft

>       It's common to use macros to avoid writing repetitive code, such as trait

Author: It would be good to explain that on refactor the ordering may be changed from:
* most specialized
* most general
to the reverse which would cause unexpected quirks later.

Author: FYI @PLeVasseur (me) you still need to do this!

Author: Handled this now!

      :id: non_compl_ex_Gb4zimei8cNI
      :status: draft

>       (example of a simple expansion using a proc-macro)

Author: ```suggestion
      Explanation of code example.
```
